### DISCUSSION TITLE
PWM doc example not working on RP2

### DISCUSSION THREAD

**Discussion by FilipSLO** [micropython/micropython#12162] [RP2040 / Pico] [answered]
I got error in the second line already from [doc example](https://docs.micropython.org/en/latest/rp2/quickref.html#pwm-pulse-width-modulation) on release 1.20.0 on RP Pico. Reproducible on older releases too.

MicroPython v1.20.0 on 2023-04-26; Raspberry Pi Pico with RP2040

```
from machine import Pin, PWM

# create PWM object from a pin and set the frequency of slice 0
# and duty cycle for channel A
pwm0 = PWM(Pin(0), freq=2000, duty_u16=32768)     # gives ERROR: **TypeError: function doesn't take keyword arguments**
pwm0 = PWM(Pin(0), 2000, 32768)                   # gives ERROR: **TypeError: function takes 1 positional arguments but 3 were given**
```

It seems to me that something is not fully ported to RP2.

Thank you for your contributions.

**Comment by robert-hh** [comment]
That is changed in the nightly builds.

**Reply by FilipSLO** [reply]
Thanks. That was really fast. 
I guess it's to early to test the fix as the [latest I could find](https://micropython.org/resources/firmware/rp2-pico-20230727-unstable-v1.20.0-327-gd14ddcbdb.uf2) it appears is from 27th of July?

**Reply by FilipSLO** [reply]
And [no commit](https://github.com/micropython/micropython/commits/master) either.

I assumed that you did something to solve this since you changed the reported issue to the discussion. 
So can you please elaborate was has changed or what I'm doing wrong?

**Reply by robert-hh** [reply]
There was PR #10850 focused on PWM. In this PR, keyword arguments were added to the PWM module of the RP2 port. It was merged on May 4th. So the doc example works now and matches the actual state of the nightly builds.

**Comment by scruss** [comment, marked as answer]
Also another reason to note the warning at the top of the *latest* documentation page:

> This is the documentation for the latest development branch of MicroPython and may refer to features that are not available in released versions. If you are looking for the documentation for a specific release, use the drop-down menu on the left and select the desired version.

The examples in the [v1.20.0 RP2 PWM docs](https://docs.micropython.org/en/v1.20.0/rp2/quickref.html#pwm-pulse-width-modulation) work as expected with v1.2.0:

```python3
from machine import Pin, PWM

pwm0 = PWM(Pin(0))      # create PWM object from a pin
pwm0.freq()             # get current frequency
pwm0.freq(1000)         # set frequency
pwm0.duty_u16()         # get current duty cycle, range 0-65535
pwm0.duty_u16(200)      # set duty cycle, range 0-65535
pwm0.deinit()           # turn off PWM on the pin
```

The examples from the [latest RP2 PWM docs](https://docs.micropython.org/en/latest/rp2/quickref.html#pwm-pulse-width-modulation), however, fail with v1.20.0:

```python3
from machine import Pin, PWM

# create PWM object from a pin and set the frequency of slice 0
# and duty cycle for channel A
pwm0 = PWM(Pin(0), freq=2000, duty_u16=32768)
  …
```

The *latest* docs may contain features not implemented in the most recent release, so please check the docs version (in the "Versions and downloads" tab) against the version of MicroPython you're running. Sometimes the docs are wrong, though, but this one is a version thing.

**Comment by FilipSLO** [comment]
> So the doc example works now and matches the actual state of the nightly builds.

Oh, just days after 1.20.0 release. 

Thanks to both to take time for explaining this in detail! I'm really sorry for the confusion! Will try again in a few days and close this to treat the issue properly.

**Comment by FilipSLO** [comment]
Other then case bellow it works as expected. It gives ERROR in NB **v1.20.0-327-gd14ddcbdb on 2023-07-27**: _TypeError: extra positional arguments given_.

`pwm0 = PWM(Pin(0), 2000, 32768)`

At least I understand the docs the way that this should work. Or am I wrong?

**Comment by robert-hh** [comment]
Except Pin and freq all arguments are keyword only. You cannot supply them as positional argument. And you must supply either duty_u16 or duty_ns. If you supply both, the behavior is not specified.

**Comment by FilipSLO** [comment]
Thanks for your kind, patient and informative response, now it makes much more sense.
